src/runtime/vm/vm.cc

```diff
@@ -899,8 +920,9 @@ void VirtualMachine::RunLoop(const std::vector<Index>& output_tensor_reg_indices
         ICHECK_EQ(actual_src_dev.device_type, inst_src_dev.device_type);
         ICHECK_EQ(actual_src_dev.device_id, inst_src_dev.device_id);
         Device dst_dev = GetDevice(instr.device_copy.dst_device_index);
+        auto mem_scope = exec_->virtual_devices[instr.device_copy.dst_device_index].second;
 
-        NDArray dst_data = src_data.CopyTo(dst_dev);
+        NDArray dst_data = src_data.CopyTo(dst_dev, String(mem_scope));
```

Comment:
I tried to do that, but it looks like we cannot avoid it. We cannot store `mem_scope` in `virtual_devices` as a `String` because of serialization problems. We could use `std::string` in `CopyTo` method, but it looks like in `NDArray` it is better to use `String` because of keeping uniform interface.

Reply:
Does the `instr.device_copy.dst` NDArray not have the correct memory scope already? 

1. If so, then can we use it instead of the copy here?

2. If not, then I would not expect `instr.device_copy.src` to have the correct mem_scope either. In that case you would need to construct a new ndarray from a container using the src dataptr and the correct memory scope.

Reply:
Sorry, @csullivan, probably I didn't get your question. So, I'll try to reply in details and please, correct me if I won't answer on your questions.

`device_copy` instruction contains three fields: `src register`, `src device index` and `dst device index`. So I extract `mem_scope` for dst memory from `dst device index` and it gets me the correct memory scope for destination array.

Answering on the first question: If you speak about copy data to dst NDArray, then no, we have to do such copy because dst NDArray will be created after this copy. If you speak about creating a string copy, then I have answered on this question in the previous message. 

Answering on the second question: I extract destination memory scope from `inst.device_copy.dst_device_index`, so extracted memory scope is correct. And I create a new NDArray from a src dataptr with correct memory scope.